### DISCUSSION TITLE
How to use URQL in a service to get a loading state of the query?

### DISCUSSION THREAD

**Discussion by sviat-barbutsa** [urql-graphql/urql#3089] [Q&A] [answered]
Hi everyone. I would like to kindly ask you about some help because I've stuck with a very interesting nuance. 
I use URQL inside the service (repository pattern) and I can't use the Hooks so I just directly use the URQL core API as described in docs. Then I trigger this methog from a Mobx store to save\persist\operate the data. 
Everything works like a charm however I faced a problem that in the scenario of non-using hooks I don't have access to the cool URQL features such as `loading` state.

I have read in docs about the Wonka lib but, to be honest, I didn't understand how it can be used in my setup at all because I can't  understand how to get the stream from the API through service and Mobx store in the client

Maybe someone have some ideas? I'll appreciate any help!

Here is the example of setup:

**Mutation**
```
export const JOIN_COMMUNITY = `
  mutation JoinCommunity ($communityId: String!) {
    joinCommunity(communityId: $communityId) 
  }
`;
```

**API service**
```
import {
  JOIN_COMMUNITY,
} from '../graphql/queries';

import {
  ICommunityRepository,
} from './Community.repository.types';

export class CommunityRepository implements ICommunityRepository {
  gqlClient;

  // currently it's urql
  constructor(gqlClient: Client) {
    this.gqlClient = gqlClient;
  }
  async joinCommunity(communityId: string) {
    const response = await this.gqlClient
      .mutation(JOIN_COMMUNITY, {
        communityId: communityId,
      })
      .toPromise();

    const { data, error } = response;

    const res = {
      data: data.joinCommunity,
      error: error,
      pageInfo: null,
      originalResponse: response,
    };

    return res;
  }

}
```

**Mobx Store (interactive layer)** (simplified store. Removed all other code for easier reading)
```
import { CommunityRepository } from '../repositories/Community.repository';

import { Community } from '@landing/src/types';

export class CommunityStore implements ICommunityStore<IRootStore> {
  rootStore: IRootStore;
  repository: CommunityRepository;
....
  constructor(rootStore: IRootStore, repository: CommunityRepository) {
    this.rootStore = rootStore;
    this.repository = repository;
  }

  joinCommunity = async (communityId: string) => {
    const { data, error } = await this.repository.joinCommunity(communityId);

    if (error) {
      console.error(`Community Error ->>> ${error}`);
      return null;
    }

  };
}

```

**Usage**
```
function MyComponent() {
 return (<button onClick={() => {
          communityStore.joinCommunity(id);
        }}>Join community</button>)
}
```

**Comment by kitten** [comment]
I'm only seeing a mutation here. For mutations the loading state we apply is simple: has a result been delivered? If not it's loading.

The same kind of applies to queries. But our new TSDocs will give more insights on that in the next batch of major releases: https://github.com/urql-graphql/urql/blob/c417be2083423433b75e3951ee8084ddbc3735b7/packages/react-urql/src/hooks/useQuery.ts#L95-L116

I don't know much about MobX, or at least I don't remember enough but you'd need to find a reactive/stream/update pattern to make use of multiple results for queries, to go beyond promises (which are fine for mutations of course).

Then, once you have that, these docs apply for streams: https://formidable.com/open-source/urql/docs/basics/core/#subscribing-to-results

**Reply by sviat-barbutsa** [reply]
Thank you @kitten  for your quick reaction and help. 
It's just an example with mutation but for a query I have the same situation. I really want to avoid a situation when should manually setup `loading: true | false `on every query/mutation when I use the `urqlClient.query` or `urqlClient.mutation` so that's why I was wondering if it's possible to avoid it. You gave me an idea to try smth like rx.js to catch the stream and populate this loading value automatically based on the query/mutation response! 
I'll update this post when I'll have some result

---
Update:

I'm trying to use `wonka.subscribe` as mentioned in the docs to get the `loading` phases of the query however it still produces only the final result similar like `toPromise()`

```
import { pipe, subscribe } from 'wonka';
....
    const { unsubscribe } = pipe(
      this.gqlClient.query(GET_USER_COMMUNITIES, {
        ownerId: userId,
      }),
      subscribe((result) => {
        console.log('WONKA SUBSCRIBE'); // It renders only once
        console.log(result); // { data: ... }
      })
    ); 
..
```

**Comment by sviat-barbutsa** [comment, marked as answer]
Hi guys. Here is the update of how I solved this problem.
After a lot of tryings and diving deep into URQL I found that URQL doesn't have a `loading | fetching` state in the case of using **gqlClient.query**. It just returns a stream of data. 
In the `useQuery` hook that provides **react-urql** the `loading` state handles manually inside this hook.

So if you want to use some thirdparty service/repository pattern/state management system like MOBX etc. so you can't use the `useQuery` hook etc. You need to use the URQL API of the core like `client.query` so you MUST handle it by yourself. The best way to achieve it will be to create some common data shape that will be used for all the fields/data entities you want to have with `loading` state and manually update it.

In my case it looks like: 
```
export type CommonApiDataShapeType<T> = {
  data: T;
  loading: boolean;
  error: string | any;
  meta: {
    [index: string]: any;
  };
};
```

So  when I call the appropriated method I manually set the `loading`. 
P.S.
In my case I must had use the _Subscription_ instead of promisifying the data as I use the **optimistic update** and I found that when I used the **client.query(...).toPromise()** it wasn't been populated  by the **Optimistic Response** data I had made in urql. Thus I was forced to use the subscription because **Optimistic Response** populates the subscription. But it brings another challlenges as need to remove old subscription and create the new one if the query arguments like (id \ before | after | limit ) had been changed (at least I didn't find a way how to solve it differently). Will appreciate for any help if it's possible.

Here is the code of updating `loading` state.

```
.... MOBX STORE CLASS .....
 --- a real method with the updating loading state, for example ---
  setSubscriptionToGetUserCommunitiesOnce(ownerId: string) {
    this.repository.subscribeToGetUserCommunities(
      ownerId,
      () => {
        runInAction(() => {
          this.userCommunities = {
            ...this.userCommunities,
            loading: true,
          };
        });
      },
      this.userCommunitiesSubscriber
    );
  }

  private userCommunitiesSubscriber = action(
    (response: BaseRepositoryResponse<CommunityEdge[]>) => {
      const { error, data } = response;
      if (error) {
        console.error(`user Communities Error ->>> ${error}`);
      }

      const result = data
        ? data.map((item) => new CommunityModel(item.node, this.rootStore))
        : [];

      this.userCommunities = {
        ...this.userCommunities,
        data: result,
        loading: false,
        error: error,
      };
    }
  );
```
